### PR TITLE
Elastic agent core publishing

### PR DESCRIPTION
**Pull request by pazone**
## Proposed commit message

Elastic agent core packaging and publishing pipeline. Uploads elastic agent core without dependencies to DRA on demand or by a schedule. 

Synchronised packaging target with `8.x`. 
Added `darwin-arm64` to default packaging targets 
Added core packaging pipeline

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Pipeline successfully uploads the required artifacts 
- [ ] E2e tests are passing

## Related issues

- Relates [#2644](https://github.com/elastic/ingest-dev/issues/2644)

### CODE REVIEW

**Review by DaveSys911 — APPROVED** [review]
lgtm

File: .buildkite/x-pack/elastic-agent/scripts/steps/build-agent-core.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+BEAT_VERSION=$(grep -oE '[0-9]+\.[0-9]+\.[0-9]+(\-[a-zA-Z]+[0-9]+)?' "libbeat/version/version.go")
+BEAT_VERSION_FULL=$BEAT_VERSION
+if [ "$DRA_WORKFLOW" == "snapshot" ]; then
+    SNAPSHOT="true"
+    BEAT_VERSION_FULL="${BEAT_VERSION}-SNAPSHOT"
+fi
+
+SNAPSHOT=$SNAPSHOT mage -d x-pack/elastic-agent packageAgentCore
+chmod -R 777 x-pack/elastic-agent/build/distributions
```
Comment (pchila) [inline .buildkite/x-pack/elastic-agent/scripts/steps/build-agent-core.sh:17]: Not sure why we are applying the 777 permissions recursively here, the archives we have do not need to be executable, maybe we want to split this between directory and files permissions:

```suggestion
chmod 777 x-pack/elastic-agent/build/distributions
chmod 664 x-pack/elastic-agent/build/distributions/*
```

File: x-pack/elastic-agent/magefile.go [hunk 2]
```diff
@@ -456,6 +456,19 @@ func CrossBuildGoDaemon() error {
 	return devtools.CrossBuildGoDaemon()
 }
 
+// PackageAgentCore cross-builds and packages distribution artifacts containing
+// only elastic-agent binaries with no extra files or dependencies.
+func PackageAgentCore() {
+	start := time.Now()
+	defer func() { fmt.Println("packageAgentCore ran for", time.Since(start)) }()
+
+	mg.Deps(CrossBuild, CrossBuildGoDaemon)
```
Comment (pchila) [inline x-pack/elastic-agent/magefile.go:465]: I don't think that GoDaemon is included in the core package (at least it's not in the 8.X versions) and it's needed only by the rpm or deb full agent packages so I think we can skip it here and save some time and money ;)

**Review by pchila — APPROVED** [review]
Just a couple of small improvements we can make for permission and avoid building unnecessary binaries, looks good overall

File: x-pack/elastic-agent/magefile.go (hunk 2 above)
Comment (pazone) [reply x-pack/elastic-agent/magefile.go:465]: It came from the elastic-agent main branch as it is :)

Comment (pchila) [reply x-pack/elastic-agent/magefile.go:465]: Then maybe we have to get rid of it also on the elastic-agent main... What I know for sure is that the `elastic-agent-core` DRA do not include `god` executable in the package, that's why I was thinking of getting rid of the dependency for this target ;)

I created an issue to track this https://github.com/elastic/elastic-agent/issues/4467 and check if we can get rid of this also from elastic-agent main branch :boom:

Comment (pazone) [reply x-pack/elastic-agent/magefile.go]: Done and it works :)

File: dev-tools/mage/platforms.go [hunk 3]
```diff
@@ -33,7 +33,7 @@ var BuildPlatforms = BuildPlatformList{
 	{"darwin/386", CGOSupported | CrossBuildSupported},
 	{"darwin/amd64", CGOSupported | CrossBuildSupported | Default},
 	{"darwin/arm", CGOSupported},
-	{"darwin/arm64", CGOSupported},
+	{"darwin/arm64", CGOSupported | CrossBuildSupported | Default},
```
Comment (v1v) [inline dev-tools/mage/platforms.go:36]: Here

**Comment by pazone** [comment]
BLocked by https://github.com/elastic/infra/pull/40818

**Comment by DaveSys911** [comment]
> Blocked by [elastic/infra#40818](https://github.com/elastic/infra/pull/40818)

Approved & queued for merge

**Comment by cmacknz** [comment]
This is a late comment, but `elastic-agent-core` isn't used in 7.17, it exists purely to support the independent agent releases project which we aren't planning to support in 7.17.

The `mage package` target doesn't consume this artifact at all anywhere right now, the elastic-agent repository will only begin to use it once https://github.com/elastic/elastic-agent/pull/4403 is merged.

File: .buildkite/x-pack/elastic-agent/.buildkite/pipeline.xpack.elastic-agent-binary-dra.yml [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
```
Comment (v1v) [inline .buildkite/x-pack/elastic-agent/.buildkite/pipeline.xpack.elastic-agent-binary-dra.yml:1]: What's the reason for using this path that contains `.buildkite` twice (parent and sibling)? As far as I see, there are already pipelines in the folder `.buildkite/x-pack/elastic-agent`:
- https://github.com/elastic/beats/tree/7.17/.buildkite/x-pack/elastic-agent

**Pull request merged**